volume/manager.go

```diff
@@ -0,0 +1,147 @@
+package volume
+
+import (
+	"context"
+	liberr "github.com/konveyor/controller/pkg/error"
+	"github.com/konveyor/controller/pkg/logging"
+	"github.com/konveyor/tackle2-hub/api"
+	"github.com/konveyor/tackle2-hub/model"
+	"github.com/konveyor/tackle2-hub/settings"
+	"github.com/konveyor/tackle2-hub/task"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+	k8s "sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+const (
+	Unit = time.Minute
+)
+
+var (
+	Settings = &settings.Settings
+	Log      = logging.WithName("volume-scheduler")
+	TTL      = 2 * Settings.Frequency.Volume
+)
+
+//
+// Manager provides task management.
+type Manager struct {
+	// DB
+	DB *gorm.DB
+	// k8s client.
+	Client k8s.Client
+	// taskID
+	taskID uint
+}
+
+//
+// Run the manager.
+func (m *Manager) Run(ctx context.Context) {
+	go func() {
+		Log.Info("Started.")
+		defer func() {
+			Log.Info("Died.")
+		}()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				m.update()
+				m.pause()
+			}
+		}
+	}()
+}
+
+//
+// Pause.
+func (m *Manager) pause() {
+	d := Unit * time.Duration(Settings.Frequency.Volume)
+	time.Sleep(d)
+}
+
+//
+// update volumes.
+// The last task is deleted and the task ID reused.
+func (m *Manager) update() {
+	list := []model.Volume{}
+	err := m.DB.Find(&list).Error
+	if err != nil {
+		Log.Trace(err)
+		return
+	}
+	ids := []uint{}
+	for _, v := range list {
+		ids = append(
+			ids, v.ID)
+	}
+	r := api.Task{}
+	r.Variant = "mount:report"
+	r.Name = r.Variant
+	r.Locator = r.Variant
+	r.Addon = "admin"
+	r.State = task.Ready
+	r.Priority = 1
+	r.TTL = &api.TTL{
+		Pending:   TTL,
+		Succeeded: TTL,
+		Failed:    TTL,
+	}
+	r.Data = map[string]interface{}{
+		"volumes": ids,
+	}
+	t := r.Model()
+	err = m.endTask()
+	if err != nil {
+		Log.Trace(err)
+		return
+	}
+	t.ID = m.taskID
+	err = m.DB.Create(t).Error
+	if err != nil {
+		Log.Trace(err)
+		m.taskID = 0
+	} else {
+		m.taskID = t.ID
+	}
+}
+
+//
+// endTask ends the current task.
+func (m *Manager) endTask() (err error) {
+	defer func() {
+		if err != nil {
+			m.taskID = 0
+		}
+	}()
+	if m.taskID == 0 {
+		return
+	}
+	current := &model.Task{}
+	current.ID = m.taskID
+	err = m.DB.First(current).Error
+	if err != nil {
+		return
+	}
+	rt := task.Task{Task: current}
```

Comment:
I'm probably just confused with renaming, but the ```task.Task{}``` is not the ```Task``` struct from addon/task.go (where I don't see the ```Task``` attribute)?

Reply:
It is the _Runtime_ Task from the _task_ (manager) package.  The _Runtime_ task is an object that represents the Task as it interfaces with k8s.

Reply:
Ah, thanks for explanation :+1: